BIMM185_Project
===============

Title: Differential Expression analysis between miRNA targets and mRNA data reveals non-linear 

correlation of gene expression changes in TCGA cancer types.


Abstract

It is well-known that miRNA is in charge of mRNA expression regulation, both negatively and positively. 

However, the correlation between both expression profiles has yet to be analyzed in data of various 

cancer types available on The Cancer Genome Atlas (TCGA). This study seeks to find correlation between 

fold changes of miRNA and mRNA data in breast invasive carcinoma (BRCA), kidney chromophobe 

(KICH), lung adenocarcinoma (LUAD), liver hepatocellular (LIHC), and head and neck squamous cell 

carcinoma (HNSC). Experimentally validated miRNA targets were found in order to produce more 

reliable data to be compared with mRNA datasets. The massive amount of data offered in large 

databases such as TCGA may be overwhelming for researchers without bioinformatics background; it is a 

necessity to provide user-friendly analysis such as this study to enable differential expression and 

correlation between different cancer datasets. Despite the need for such analysis, nothing similar was 

found available which stresses the significance of the discovered results. Based on the conducted 

analysis on each cancer type, using Pearson correlation, it was found that miRNA regulation in tumor 

cells is highly correlated with transcriptome wide gene expression changes, only in a few differentially 

expressed genes. These results suggest that miRNA regulation of mRNA expression does not fit the 

Pearson model and therefore does not have linear relationship. The analysis products set a base for 

further studies which would allow to draw interesting conclusions about gene expression in different 

cancer types as well as miRNA regulation.
